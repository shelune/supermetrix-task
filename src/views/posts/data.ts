import { useParams } from "react-router-dom";
import axios from "axios";
import { ComponentProps, useCallback, useEffect, useState } from "react";
import type { ApiErrorResponse, ApiPostResponse, Post } from "../../shared/api";
import type { PostsView } from "./posts";

import { postMapper } from "../../shared/utils/mapper";
import { getCookie } from "../../shared/utils/cookie";

export const MAX_PAGE = 10;

export const useData = (): ComponentProps<typeof PostsView> => {
  const [posts, setPosts] = useState<Post[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [page, setPage] = useState(1);
  const [activeUser, setActiveUser] = useState("");
  const params = useParams();

  const getData = useCallback(async () => {
    setLoading(true);
    try {
      const response = await axios.get<ApiPostResponse>(
        "https://api.supermetrics.com/assignment/posts",
        {
          params: {
            sl_token: getCookie("sl_token"),
            page,
          },
        },
      );
      const mapped = response.data.data.posts.map(postMapper);
      setPosts(posts.concat(mapped));
      setError("");
    } catch (err) {
      if (axios.isAxiosError(err) && err.response?.data) {
        const errorResponse = err.response.data as ApiErrorResponse;
        setError(errorResponse.error.message);
      } else {
        // eslint-disable-next-line no-console
        console.log("unknown error when getting posts", err);
      }
    }
    setLoading(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [page]);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    getData();
  }, [getData, page]);

  useEffect(() => {
    if (params.userId) {
      setActiveUser(params.userId);
    } else {
      setActiveUser("");
    }
  }, [params.userId]);

  const loadMore = useCallback(() => {
    if (page + 1 > MAX_PAGE) {
      return;
    }
    setPage(page + 1);
  }, [page]);

  return {
    posts,
    currentPage: page,
    activeUser,
    error,
    loading,
    loadMore,
  };
};
