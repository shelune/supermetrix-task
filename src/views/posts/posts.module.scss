@import "../../styles/utils/_index.scss";

.container {
  padding: 24px;
  display: grid;
  grid-template-columns: 1fr 2fr;
  gap: 16px;
  align-items: start;
}

.error {
  color: $red-error;
  text-align: center;
}
