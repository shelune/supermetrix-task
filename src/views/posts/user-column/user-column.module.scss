@import "../../../styles/utils/_index.scss";

.container {
  position: sticky;
  top: 16px;
}

.searchBar {
  margin-bottom: 16px;
}

.textInput {
  @include rounded-4;
  width: 75%;
  max-width: 100%;
  padding: 12px;
  border: 1px solid $gray;
}

.userList {
  max-height: 80vh;
  overflow: auto;
}

.userItem {
  @include rounded-4;
  display: flex;
  justify-content: space-between;
  padding: 16px;
  background-color: $gray-light;
  margin-bottom: 8px;
  cursor: pointer;
  border: 2px solid transparent;
  color: inherit;
  text-decoration: none;

  &.isActive {
    border-color: $gray;
  }
}

.postCount {
  @include text-subtitle-md;
  @include rounded-full;
  width: 24px;
  height: 24px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  background-color: $gray;
  color: $white;
}
