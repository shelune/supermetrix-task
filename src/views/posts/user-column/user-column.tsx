import classNames from "classnames";
import React, { FC, useMemo, useState } from "react";
import { Link } from "react-router-dom";
import type { Post } from "../../../shared/api";
import type { UserData } from "../posts";

import css from "./user-column.module.scss";

type Props = {
  posts: Post[];
  activeUser: string;
  loading: boolean;
};

export const UserColumn: FC<Props> = ({ posts, activeUser, loading }) => {
  const [search, setSearch] = useState("");
  const users: UserData[] = useMemo(() => {
    const dupedUsers = posts.map((post) => ({
      name: post.authorName,
      userId: post.authorId,
    }));
    const filteredUsers = dupedUsers.filter(
      (user, index, array) =>
        array.findIndex((match) => match.userId === user.userId) === index,
    );
    return filteredUsers.map((user) => ({
      ...user,
      postCount: posts.filter((post) => post.authorId === user.userId).length,
    }));
  }, [posts]);

  const filteredUsers = useMemo(
    () =>
      users.filter((user) =>
        user.name.toLowerCase().includes(search.toLowerCase()),
      ),
    [search, users],
  );
  return (
    <div className={css.container}>
      <div className={css.columnHeader}>
        <div className={css.searchBar}>
          <input
            data-testid="user-search-text"
            type="text"
            className={css.textInput}
            value={search}
            onChange={(e) => setSearch(e.target.value)}
            placeholder="Search"
          />
        </div>
      </div>
      <div className={css.columnBody} data-testid="user-column-body">
        {loading ? (
          <div className={css.userListLoading}>Loading...</div>
        ) : (
          <div className={css.userList}>
            {filteredUsers.map((user) =>
              activeUser === user.userId ? (
                <Link
                  data-testid="user-item"
                  to="/posts"
                  key={user.userId}
                  className={classNames(css.userItem, css.isActive)}
                >
                  <div className={css.name}>{user.name}</div>
                  <div className={css.postCount}>{user.postCount}</div>
                </Link>
              ) : (
                <Link
                  data-testid="user-item"
                  to={`/posts/${user.userId}`}
                  key={user.userId}
                  className={classNames(css.userItem)}
                >
                  <div className={css.name}>{user.name}</div>
                  <div className={css.postCount}>{user.postCount}</div>
                </Link>
              ),
            )}
          </div>
        )}
      </div>
    </div>
  );
};
