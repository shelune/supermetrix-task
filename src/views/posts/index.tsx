import React, { FC, memo } from "react";
import { useData } from "./data";
import { PostsView as MainComponent } from "./posts";

export const PostsView: FC = memo(
  () => {
    const props = useData();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <MainComponent {...props} />;
  },
  () => true,
);
