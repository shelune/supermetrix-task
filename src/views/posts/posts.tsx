import React, { FC } from "react";
import { Post } from "../../shared/api";
import { PostColumn } from "./post-column/post-column";

import css from "./posts.module.scss";
import { UserColumn } from "./user-column/user-column";

export type UserData = {
  name: string;
  userId: string;
  postCount: number;
};

type Props = {
  posts: Post[];
  currentPage: number;
  activeUser: string;
  error: string;
  loading: boolean;
  loadMore: () => void;
};

export const PostsView: FC<Props> = ({
  posts,
  currentPage,
  activeUser,
  error,
  loading,
  loadMore,
}) => (
  <>
    {error && <div className={css.error}>{error}</div>}
    <div className={css.container}>
      <UserColumn posts={posts} activeUser={activeUser} loading={loading} />
      <PostColumn
        posts={posts}
        currentPage={currentPage}
        loading={loading}
        loadMore={loadMore}
      />
    </div>
  </>
);
