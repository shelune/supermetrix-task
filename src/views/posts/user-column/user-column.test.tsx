import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import React, { ReactNode } from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { UserColumn } from "./user-column";
import { mockPosts } from "../../../shared/__mock__/posts";

function renderWithRouter(children: ReactNode) {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {children}
    </Router>,
  );
}

describe("UserColumn", () => {
  it("should render", () => {
    const { container } = renderWithRouter(
      <UserColumn posts={mockPosts} activeUser="" loading={false} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should have correct numbers of users based on posts", () => {
    const { getAllByTestId } = renderWithRouter(
      <UserColumn posts={mockPosts} activeUser="" loading={false} />,
    );
    expect(getAllByTestId("user-item")).toHaveLength(20);
  });

  it("should filter out users based on search", () => {
    const { getByTestId, getAllByTestId, getByText } = renderWithRouter(
      <UserColumn posts={mockPosts} activeUser="" loading={false} />,
    );
    const userSearch = getByTestId("user-search-text");
    act(() => {
      fireEvent.change(userSearch, {
        target: {
          value: "Quyen",
        },
      });
    });
    // eslint-disable-next-line jest-dom/prefer-in-document
    expect(getAllByTestId("user-item")).toHaveLength(1);
    expect(getByText("Quyen Pellegrini")).toBeInTheDocument();
  });

  it("should render accordingly when there's an active user", () => {
    const { container } = renderWithRouter(
      <UserColumn posts={mockPosts} activeUser="user_19" loading={false} />,
    );
    expect(container.getElementsByClassName("isActive")).toHaveLength(1);
  });

  it("should render loading state", () => {
    const { container } = renderWithRouter(
      <UserColumn posts={mockPosts} activeUser="" loading />,
    );
    expect(container).toMatchSnapshot();
  });
});
