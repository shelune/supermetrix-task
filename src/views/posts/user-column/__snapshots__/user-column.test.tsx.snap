// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserColumn should render 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="columnHeader"
    >
      <div
        class="searchBar"
      >
        <input
          class="textInput"
          data-testid="user-search-text"
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="columnBody"
      data-testid="user-column-body"
    >
      <div
        class="userList"
      >
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_15"
        >
          <div
            class="name"
          >
            Yolande Urrutia
          </div>
          <div
            class="postCount"
          >
            7
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_14"
        >
          <div
            class="name"
          >
            Woodrow Lindholm
          </div>
          <div
            class="postCount"
          >
            7
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_7"
        >
          <div
            class="name"
          >
            Gigi Richter
          </div>
          <div
            class="postCount"
          >
            3
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_10"
        >
          <div
            class="name"
          >
            Rafael Althoff
          </div>
          <div
            class="postCount"
          >
            9
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_12"
        >
          <div
            class="name"
          >
            Lashanda Small
          </div>
          <div
            class="postCount"
          >
            11
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_19"
        >
          <div
            class="name"
          >
            Quyen Pellegrini
          </div>
          <div
            class="postCount"
          >
            4
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_6"
        >
          <div
            class="name"
          >
            Carly Alvarez
          </div>
          <div
            class="postCount"
          >
            7
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_17"
        >
          <div
            class="name"
          >
            Mandie Nagao
          </div>
          <div
            class="postCount"
          >
            8
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_13"
        >
          <div
            class="name"
          >
            Regenia Boice
          </div>
          <div
            class="postCount"
          >
            7
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_2"
        >
          <div
            class="name"
          >
            Nydia Croff
          </div>
          <div
            class="postCount"
          >
            5
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_8"
        >
          <div
            class="name"
          >
            Maxie Marceau
          </div>
          <div
            class="postCount"
          >
            1
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_18"
        >
          <div
            class="name"
          >
            Ethelene Maggi
          </div>
          <div
            class="postCount"
          >
            4
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_3"
        >
          <div
            class="name"
          >
            Leonarda Schult
          </div>
          <div
            class="postCount"
          >
            2
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_4"
        >
          <div
            class="name"
          >
            Britany Heise
          </div>
          <div
            class="postCount"
          >
            6
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_16"
        >
          <div
            class="name"
          >
            Isidro Schuett
          </div>
          <div
            class="postCount"
          >
            3
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_1"
        >
          <div
            class="name"
          >
            Filomena Cort
          </div>
          <div
            class="postCount"
          >
            4
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_9"
        >
          <div
            class="name"
          >
            Rosann Eide
          </div>
          <div
            class="postCount"
          >
            2
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_11"
        >
          <div
            class="name"
          >
            Macie Mckamey
          </div>
          <div
            class="postCount"
          >
            5
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_0"
        >
          <div
            class="name"
          >
            Lael Vassel
          </div>
          <div
            class="postCount"
          >
            2
          </div>
        </a>
        <a
          class="userItem"
          data-testid="user-item"
          href="/posts/user_5"
        >
          <div
            class="name"
          >
            Carson Smithson
          </div>
          <div
            class="postCount"
          >
            3
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`UserColumn should render loading state 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="columnHeader"
    >
      <div
        class="searchBar"
      >
        <input
          class="textInput"
          data-testid="user-search-text"
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="columnBody"
      data-testid="user-column-body"
    >
      <div
        class="userListLoading"
      >
        Loading...
      </div>
    </div>
  </div>
</div>
`;
